crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs

```diff
@@ -0,0 +1,403 @@
+use crate::{context::LintContext, rule::Rule, AstNode};
+use oxc_ast::{
+    ast::{Argument, CallExpression, Expression, VariableDeclarationKind},
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+
+fn warn() -> OxcDiagnostic {
+    OxcDiagnostic::warn(
+        "eslint-plugin-unicorn(no-useless-undefined): Do not use useless `undefined`.",
+    )
+    .with_help("Consider using `null` instead.")
+}
+fn no_useless_undefined_diagnostic(span0: Span) -> OxcDiagnostic {
+    warn().with_label(span0)
+}
+fn no_useless_undefined_diagnostic_spans(spans: Vec<Span>) -> OxcDiagnostic {
+    warn().with_labels(spans)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUselessUndefined {
+    check_arguments: bool,
+    check_arrow_function_body: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Do not use useless `undefined`.
+    ///
+    /// ### Why is this bad?
+    /// `undefined` is the default value for new variables, parameters, return statements, etc… so specifying it doesn't make any difference.
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // bad
+    /// let foo = undefined;
+    /// // good:
+    /// let foo;
+    /// ```
+    NoUselessUndefined,
+    pedantic,
+);
+
+const COMPARE_FUNCTION_NAMES: &[&str] = &[
+    "is",
+    "equal",
+    "notEqual",
+    "strictEqual",
+    "notStrictEqual",
+    "propertyVal",
+    "notPropertyVal",
+    "not",
+    "include",
+    "property",
+    "toBe",
+    "toHaveBeenCalledWith",
+    "toContain",
+    "toContainEqual",
+    "toEqual",
+    "same",
+    "notSame",
+    "strictSame",
+    "strictNotSame",
+];
+
+fn is_match_ignore_func_name(name: &str) -> bool {
+    COMPARE_FUNCTION_NAMES.contains(&name)
+        // `array.push(undefined)`
+        || name == "push"
```

Comment:
Create a static set for these names. Also, consider sorting them and using a binary search for faster checks.

Reply:
I attempted to use phf::phf_set! like this file, but I'm confused why using set still requires binary search.
https://github.com/oxc-project/oxc/blob/main/crates/oxc_linter/src/rules/unicorn/prefer_add_event_listener.rs#L68